pkg/controllers/corecluster/corecluster_controller.go

```diff
@@ -0,0 +1,243 @@
+	capiInfraClusterAPIVersionV1Beta1 = "infrastructure.cluster.x-k8s.io/v1beta1"
+	capiInfraClusterAPIVersionV1Beta2 = "infrastructure.cluster.x-k8s.io/v1beta2"
+	clusterOperatorName               = "cluster-api"
+)
+
+var (
+	errPlatformStatusShouldNotBeNil                   = errors.New("infrastructure PlatformStatus should not be nil")
+	errMappingOCPPlatformToInfraClusterKindAndVersion = errors.New("failed to map OpenShift Infrastructure PlatformType to a Cluster API infra cluster kind, platform not supported")
+	errOCPInfraShouldNotBeNil                         = errors.New("ocp infrastructure object should not be nil")
+	errOCPInfrastructureNameShouldNotBeEmpty          = errors.New("ocp infrastructure object's status.infrastructureName should not be empty")
+)
+
+// CoreClusterReconciler reconciles a Cluster object.
+type CoreClusterReconciler struct {
+	operatorstatus.ClusterOperatorStatusClient
+	Cluster  *clusterv1.Cluster
+	Infra    *configv1.Infrastructure
+	Platform configv1.PlatformType
+}
+
+// SetupWithManager sets the CoreClusterReconciler controller up with the given manager.
+func (r *CoreClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := ctrl.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		For(&configv1.ClusterOperator{}, builder.WithPredicates(clusterOperatorPredicates())).
+		Watches(
+			&clusterv1.Cluster{},
+			handler.EnqueueRequestsFromMapFunc(toClusterOperator),
+			builder.WithPredicates(coreClusterPredicate(r.ManagedNamespace)),
+		).
+		Complete(r); err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the core cluster object for the openshift-cluster-api namespace.
+func (r *CoreClusterReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithName(controllerName)
+	logger.V(1).Info("Reconciling core cluster")
+
+	ocpInfrastructureName, err := getOCPInfrastructureName(r.Infra)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to obtain ocp infrastructure name: %w", err)
+	}
+
+	cluster, err := r.ensureCoreCluster(ctx, client.ObjectKey{Namespace: req.Namespace, Name: ocpInfrastructureName}, logger)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to ensure new core cluster: %w", err)
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		if err := r.SetStatusAvailable(ctx, ""); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.ensureCoreClusterControlPlaneInitializedCondition(ctx, cluster); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to ensure core cluster has the ControlPlaneInitializedCondition: %w", err)
+	}
+
+	if err := r.SetStatusAvailable(ctx, ""); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+	}
+
+	defer logger.V(1).Info("Finished reconciling core cluster")
```

Comment:
A defer at the end here is kinda weird, should it not be immediately after the start log on line 86?